core/src/test/java/org/apache/iceberg/TestContentFileParser.java

```diff
@@ -0,0 +1,245 @@
+        Arguments.of(
+            PartitionSpec.unpartitioned(),
+            deleteFileWithRequiredOnly(PartitionSpec.unpartitioned())),
+        Arguments.of(
+            PartitionSpec.unpartitioned(),
+            deleteFileWithAllOptional(PartitionSpec.unpartitioned())),
+        Arguments.of(TableTestBase.SPEC, deleteFileWithRequiredOnly(TableTestBase.SPEC)),
+        Arguments.of(TableTestBase.SPEC, deleteFileWithAllOptional(TableTestBase.SPEC)));
+  }
+
+  private static DeleteFile deleteFileWithRequiredOnly(PartitionSpec spec) {
+    PartitionData partitionData = null;
+    if (spec.isPartitioned()) {
+      partitionData = new PartitionData(spec.partitionType());
+      partitionData.set(0, 9);
+    }
+
+    return new GenericDeleteFile(
+        spec.specId(),
+        FileContent.POSITION_DELETES,
+        "/path/to/delete-a.parquet",
+        FileFormat.PARQUET,
+        partitionData,
+        1234,
+        new Metrics(9L, null, null, null, null),
+        null,
+        null,
+        null);
+  }
+
+  private static DeleteFile deleteFileWithAllOptional(PartitionSpec spec) {
+    PartitionData partitionData = new PartitionData(spec.partitionType());
+    if (spec.isPartitioned()) {
+      partitionData.set(0, 9);
+    }
+
+    Metrics metrics =
+        new Metrics(
+            10L, // record count
+            ImmutableMap.of(3, 100L, 4, 200L), // column sizes
+            ImmutableMap.of(3, 90L, 4, 180L), // value counts
+            ImmutableMap.of(3, 10L, 4, 20L), // null value counts
+            ImmutableMap.of(3, 0L, 4, 0L), // nan value counts
+            ImmutableMap.of(
+                3,
+                Conversions.toByteBuffer(Types.IntegerType.get(), 1),
+                4,
+                Conversions.toByteBuffer(Types.IntegerType.get(), 2)), // lower bounds
+            ImmutableMap.of(
+                3,
+                Conversions.toByteBuffer(Types.IntegerType.get(), 5),
+                4,
+                Conversions.toByteBuffer(Types.IntegerType.get(), 10)) // upperbounds
+            );
+
+    return new GenericDeleteFile(
+        spec.specId(),
+        FileContent.EQUALITY_DELETES,
+        "/path/to/delete-with-stats.parquet",
+        FileFormat.PARQUET,
+        partitionData,
+        1234,
+        metrics,
+        new int[] {3},
+        1,
+        ByteBuffer.wrap(new byte[16]));
+  }
+
+  @ParameterizedTest
+  @MethodSource("provideSpecAndDataFile")
+  public void testDataFile(PartitionSpec spec, DataFile dataFile) throws Exception {
+    StringWriter stringWriter = new StringWriter();
+    JsonGenerator jsonStringGenerator = JsonUtil.factory().createGenerator(stringWriter);
+
+    ContentFileParser.toJson(dataFile, spec, jsonStringGenerator);
+    jsonStringGenerator.close();
+    String jsonStr = stringWriter.toString();
+
+    JsonNode jsonNode = JsonUtil.mapper().readTree(jsonStr);
+    DataFile deserializedFile = (DataFile) ContentFileParser.fromJson(jsonNode, spec);
```

Comment:
rather than casting I think we should check that it's actually a DataFile:

```
ContentFile<?> contentFile = ContentFileParser.fromJson(jsonNode, spec);
Assertions.assertThat(contentFile).isInstanceOf(DataFile.class);
```